src/wpantund/NCPInstanceBase-Addresses.cpp

```diff
@@ -30,122 +30,529 @@
 #include <algorithm>
 #include "socket-utils.h"
 #include "SuperSocket.h"
+#include "IPv6Helpers.h"
 
 using namespace nl;
 using namespace wpantund;
 
+std::string
+NCPInstanceBase::EntryBase::get_origin_as_string(void) const
+{
+	const char *ret = "unknown";
+
+	switch (mOrigin) {
+	case kOriginThreadNCP:
+		ret =  "ncp";
+		break;
+
+	case kOriginPrimaryInterface:
+		ret =  "intface";
+		break;
+
+	case kOriginUser:
+		ret = "user";
+		break;
+	}
+
+	return ret;
+}
+
+NCPInstanceBase::UnicastAddressEntry::UnicastAddressEntry(
+	Origin origin,
+	uint8_t prefix_len,
+	bool filtered,
+	uint32_t valid_lifetime,
+	uint32_t preferred_lifetime
+) :	EntryBase(origin)
+{
+	mPrefixLen = prefix_len;
+	mFiltered = filtered;
+	set_valid_lifetime(valid_lifetime);
+	set_preferred_lifetime(preferred_lifetime);
+}
+
+void
+NCPInstanceBase::UnicastAddressEntry::set_valid_lifetime(uint32_t valid_lifetime)
+{
+	mValidLifetime = valid_lifetime;
+
+	mValidLifetimeExpiration = (valid_lifetime == UINT32_MAX)
+		? TIME_DISTANT_FUTURE
+		: time_get_monotonic() + valid_lifetime;
```

Comment:
This is tricky, because last time I checked this sort of thing wasn't happening on the NCP. What action do we take when this expires? I would think the NCP would be in charge of expiring things.